pinot-core/src/main/java/org/apache/pinot/core/segment/memory/PinotNonNativeUnsafeByteBuffer.java

```diff
@@ -0,0 +1,355 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.memory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.pinot.core.util.CleanerUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.misc.Unsafe;
+import sun.nio.ch.DirectBuffer;
+
+
+@ThreadSafe
+public class PinotNonNativeUnsafeByteBuffer extends PinotDataBuffer {
+  private final DirectBuffer _buffer;
+  private final long _address;
+  private final boolean _flushable;
+  private static final Unsafe UNSAFE = getUnsafe();
+
+  private static Unsafe getUnsafe() {
+    try {
+      Field var0 = Unsafe.class.getDeclaredField("theUnsafe");
+      var0.setAccessible(true);
+      return (Unsafe)(var0.get((Object)null));
+    } catch (NoSuchFieldException e) {
+      throw new IllegalStateException("sun.misc.Unsafe is not available in this JVM", e);
+    } catch (IllegalAccessException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public static PinotNonNativeUnsafeByteBuffer allocateDirect(int size) {
+    return new PinotNonNativeUnsafeByteBuffer(ByteBuffer.allocateDirect(size), true, false);
+  }
+
+  static PinotNonNativeUnsafeByteBuffer loadFile(File file, long offset, int size)
+      throws IOException {
+    PinotNonNativeUnsafeByteBuffer buffer = allocateDirect(size);
+    buffer.readFrom(0, file, offset, size);
+    return buffer;
+  }
+
+  static PinotNonNativeUnsafeByteBuffer mapFile(File file, boolean readOnly, long offset, int size)
+      throws IOException {
+    if (readOnly) {
+      try (FileChannel fileChannel = new RandomAccessFile(file, "r").getChannel()) {
+        ByteBuffer buffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, offset, size);
+        return new PinotNonNativeUnsafeByteBuffer(buffer, true, false);
+      }
+    } else {
+      try (FileChannel fileChannel = new RandomAccessFile(file, "rw").getChannel()) {
+        ByteBuffer buffer = fileChannel.map(FileChannel.MapMode.READ_WRITE, offset, size);
+        return new PinotNonNativeUnsafeByteBuffer(buffer, true, true);
+      }
+    }
+  }
+
+  private PinotNonNativeUnsafeByteBuffer(ByteBuffer buffer, boolean closeable, boolean flushable) {
+    super(closeable);
+    _buffer = (DirectBuffer)buffer;
+    _address = _buffer.address();
+    _flushable = flushable;
+  }
+
+  @Override
+  public byte getByte(int offset) {
```

Comment:
Some of these api's seem common in both classes, may be push up to a base class?